File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/wrapper/ChannelWrapper.java
package org.kevoree.library.java.wrapper;

import org.kevoree.api.Callback;
import org.kevoree.api.ChannelDispatch;
import org.kevoree.log.Log;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

/**
 *
 * Created by duke on 9/26/14.
 */
public class ChannelWrapper extends KInstanceWrapper {

    private ArrayList<StoredCall> pending = new ArrayList<StoredCall>();
    private ChannelWrapperContext context;

    @Override
    public void startInstance() throws InvocationTargetException {
        super.startInstance();
        processPending();
    }

    public void setContext(ChannelWrapperContext context) {
        this.context = context;
    }

    public ChannelWrapperContext getContext() {
        return this.context;
    }

    public void call(org.kevoree.api.Callback callback, String payload) {
        Set<String> portPaths = new HashSet<>();
        portPaths.addAll(context.getBoundPorts().keySet());
        portPaths.addAll(context.getRemotePortPaths());
        String connectedInputs = portPaths.stream().reduce((p, n) -> p + ", " + n).orElse("");

        if (isStarted()) {
            ChannelDispatch channel = (ChannelDispatch) getTargetObj();
            Log.debug(" {} -> {} -> [{}] (dispatch)", getModelElement().getName(), payload, connectedInputs);
            try {
                ClassLoader nodeCL = Thread.currentThread().getContextClassLoader();
                Thread.currentThread().setContextClassLoader(channel.getClass().getClassLoader());
                channel.dispatch(payload, callback);
                Thread.currentThread().setContextClassLoader(nodeCL);
            } catch (Throwable e) {
                Log.error("Channel \"{}\" dispatch threw an exception", e.getCause(), getModelElement().getName());
            }
        } else {
            Log.debug(" {} -> {} -> [{}] (queued)", getModelElement().getName(), payload, connectedInputs);
            pending.add(new StoredCall(payload, callback));
        }
    }

    private void processPending() {
        if (!pending.isEmpty()) {
            Thread t = new Thread(() -> {
                for (StoredCall c : pending) {
                    call(c.getCallback(), c.getPayload());
                }
                pending.clear();
            });
            t.start();
        }
    }

    private class StoredCall {

        private String payload;
        private Callback callback;

        private StoredCall(String payload, Callback callback) {
            this.payload = payload;
            this.callback = callback;
        }

        private String getPayload() {
            return payload;
        }

        private void setPayload(String payload) {
            this.payload = payload;
        }

        private Callback getCallback() {
            return callback;
        }

        private void setCallback(Callback callback) {
            this.callback = callback;
        }
    }
}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/defaultNodeTypes/wrapper/ChannelWrapper.kt
package org.kevoree.library.defaultNodeTypes.wrapper;

import org.kevoree.library.defaultNodeTypes.reflect.MethodAnnotationResolver
import org.kevoree.library.defaultNodeTypes.reflect.FieldAnnotationResolver
import java.lang.reflect.InvocationTargetException
import org.kevoree.api.BootstrapService
import org.kevoree.api.ChannelContext
import org.kevoree.api.ChannelDispatch
import org.kevoree.ContainerRoot
import org.kevoree.log.Log
import org.kevoree.Channel
import org.kevoree.api.ModelService

public class ChannelWrapper(val modelElement: Channel, override val targetObj: Any, val _nodeName: String, override var tg: ThreadGroup, override val bs: BootstrapService, val modelService : ModelService) : KInstanceWrapper {

    val context: ChannelWrapperContext

    {
        context = ChannelWrapperContext(modelElement.path()!!,_nodeName,modelService)
        bs.injectService(javaClass<ChannelContext>(), context, targetObj)
    }

    override var isStarted: Boolean = false
    override val resolver: MethodAnnotationResolver = MethodAnnotationResolver(targetObj.javaClass)
    private val fieldResolver = FieldAnnotationResolver(targetObj.javaClass);

    fun call(callback: org.kevoree.api.Callback<out Any?>?, payload: Any?) {
        if(isStarted){
            (targetObj as ChannelDispatch).dispatch(payload,callback)
        }
    }

    override fun kInstanceStart(tmodel: ContainerRoot): Boolean {
        if (!isStarted) {
            try {
                val met = resolver.resolve(javaClass<org.kevoree.annotation.Start>())
                met?.invoke(targetObj)
                isStarted = true
                return true
            }catch(e: InvocationTargetException){
                Log.error("Kevoree Channel Instance Start Error !", e.getCause())
                return false
            }catch(e: Exception) {
                Log.error("Kevoree Channel Instance Start Error !", e)
                return false
            }
        } else {
            Log.error("Try to start the channel {} while it is already start", modelElement.name)
            return false
        }
    }

    override fun kInstanceStop(tmodel: ContainerRoot): Boolean {
        if (isStarted) {
            try {
                val met = resolver.resolve(javaClass<org.kevoree.annotation.Stop>())
                met?.invoke(targetObj)
                isStarted = false
                return true
            }catch(e: InvocationTargetException){
                Log.error("Kevoree Channel Instance Stop Error !", e.getCause())
                return false
            } catch(e: Exception) {
                Log.error("Kevoree Channel Instance Stop Error !", e)
                return false
            }
        } else {
            return true
        }
    }


}





